yarn-project/world-state/src/world-state-db/merkle_trees.ts

```diff
@@ -332,6 +385,282 @@ export class MerkleTrees implements MerkleTreeDb {
+   *    Why cant we?
+   *      - Index 0 has a val 0 and nextVal of 2. This is NOT enough to prove non inclusion of 2.
+   *      - Our existing tree is in a state where we cannot prove non inclusion of 3.
+   *    We do not provide a non inclusion proof to out circuit, but prompt it to look within the insertion subtree.
+   * 2. Update pending insertion subtree
+   * 3. Insert 3 into pending subtree
+   *
+   * (no inclusion proof provided)
+   *  index     0       1       2       3         4       5       -       -
+   *  -------------------------------------      ----------------------------
+   *  val       0       5      10      15         2       3       -       -
+   *  nextIdx   4       2       3       0         5       1       -       -
+   *  nextVal   2      10      15       0         3       5       -       -
+   *
+   * Inserting 20: (happy path)
+   * 1. Find the low nullifier (15) - provide inculsion proof
+   * 2. Update its pointers
+   * 3. Insert 20 into the pending subtree
+   *
+   *  index     0       1       2       3         4       5       6       -
+   *  -------------------------------------      ----------------------------
+   *  val       0       5      10      15         2       3      20       -
+   *  nextIdx   4       2       3       6         5       1       0       -
+   *  nextVal   2      10      15      20         3       5       0       -
+   *
+   * Inserting 19:
+   * 1. In this case we can find a low nullifier, but we are updating a low nullifier that has already been updated
+   *    We can provide an inclusion proof of this intermediate tree state.
+   * 2. Update its pointers
+   * 3. Insert 19 into the pending subtree
+   *
+   *  index     0       1       2       3         4       5       6       7
+   *  -------------------------------------      ----------------------------
+   *  val       0       5      10      15         2       3      20       19
+   *  nextIdx   4       2       3       7         5       1       0       6
+   *  nextVal   2      10      15      19         3       5       0       20
+   *
+   * Perform subtree insertion
+   *
+   *  index     0       1       2       3       4       5       6       7
+   *  ---------------------------------------------------------------------
+   *  val       0       5      10      15       2       3      20       19
+   *  nextIdx   4       2       3       7       5       1       0       6
+   *  nextVal   2      10      15      19       3       5       0       20
+   *
+   * TODO: this implementation will change once the zero value is changed from h(0,0,0). Changes incoming over the next sprint
+   * @param leaves - Values to insert into the tree
+   * @param includeUncommitted - Indicates whether to include uncommitted data.
+   * @returns The witness data for the leaves to be updated when inserting the new ones.
+   */
+  private async _performBaseRollupBatchInsertionProofs(
```

Comment:
Can this be moved down into the indexed tree implementation? This operation is specific to nullifier trees and it would also have the benefit that the operation would become atomic.